### PR TITLE
Add `range_formatter`

### PR DESCRIPTION
**Pull request by JMazurkiewicz**
* Remove `used by pair, tuple, (...)` comment from `_Fill_align_and_width_specs` - it no longer serves its purpose.
* Make `_Fill_align_and_width_specs_setter<T>::_Specs` member variable protected, so that we can create derived `_Range_specs_setter`.
* Implement `range_formatter`:
  * `range_formatter::format` can be used at runtime only if format context's output iterator is `back_insert_iterator<_Fmt_buffer<T>>`. Creation of nested `basic_format_context` requires copying format args from old `context<X>` to `context<Y>` and `X` and `Y` have to be the same type to do this. This should not be a problem anyway, since `basic_format_context` can be constructed only by the standard library.
  * Towards #2919.

### CODE REVIEW

File: tests/std/tests/P2286R8_text_formatting_range_formatter/test.cpp [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        fmt(STR("{:x:}"), Range(vals));
+        fmt(STR("{:x}"), Range(vals));
+        fmt(STR("{:ns}"), Range(vals));
+        fmt(STR("{:n?s}"), Range(vals));
+        fmt(STR("{:m}"), Range(vals));
+        fmt(STR("{:s:x}"), Range(vals));
+        fmt(STR("{:?s:x}"), Range(vals));
+        fmt(STR("{:?}"), Range(vals));
+        fmt(STR("{:?z}"), Range(vals));
+        fmt(STR("{::^20}"), Range(vals));
```
Comment (cpplearner) [inline tests/std/tests/P2286R8_text_formatting_range_formatter/test.cpp:334]: I think `{::^20}` should be a valid format string, with `^20` being the underlying spec.

Comment (JMazurkiewicz) [reply tests/std/tests/P2286R8_text_formatting_range_formatter/test.cpp:334]: Good catch, thanks! It looks like `parse` function tries to interpret second `:` as a `range-fill` character.

File: stl/inc/format [hunk 2]
```diff
@@ -4354,6 +4356,268 @@ public:
...
+    if (*_Begin == ':') {
+        _Callbacks._On_underlying_spec();
+    }
+
+    return _Begin;
+}
+
+template <class _Ty>
+concept _Pair_like_with_formatter =
+    _Is_specialization_v<_Ty, pair> || (_Is_specialization_v<_Ty, tuple> && tuple_size_v<_Ty> == 2);
```
Comment (frederick-vs-ja) [inline stl/inc/format:4443]: I think we should either reuse `_Is_two_tuple` introduced in #4116 or replace it with `_Pair_like_with_formatter`.
https://github.com/microsoft/STL/blob/8dc4faadafb52e3e0a627e046b41258032d9bc6a/stl/inc/format#L2469-L2476

File: tests/libcxx/expected_results.txt [hunk 3]
```diff
@@ -38,6 +38,10 @@ std/containers/sequences/vector.bool/vector.bool.fmt/parse.pass.cpp FAIL
 std/thread/thread.threads/thread.thread.class/thread.thread.id/parse.pass.cpp FAIL
 std/utilities/format/format.tuple/parse.pass.cpp FAIL
 
+# We should provide 'test_format_context_create' function in 'libcxx/test/support/test_format_context.h' file to enable those tests.
+std/utilities/format/format.range/format.range.formatter/set_brackets.pass.cpp FAIL
+std/utilities/format/format.range/format.range.formatter/set_separator.pass.cpp FAIL
+
```
Comment (frederick-vs-ja) [inline tests/libcxx/expected_results.txt:44]: There's a pre-existing section for `test_format_context_create` function in `libcxx/test/support/test_format_context.h`. Presumably we should merge both sections.
https://github.com/microsoft/STL/blob/8dc4faadafb52e3e0a627e046b41258032d9bc6a/tests/libcxx/expected_results.txt#L1106-L1127

File: stl/inc/format [hunk 4]
```diff
@@ -4354,6 +4356,268 @@ public:
...
+    }
+
+    template <_RANGES input_range _Range, class _FormatContext>
+    void _Format_as_string(_Range&& _Rng, _FormatContext& _Ctx, bool _Debug) const {
+        if constexpr (_RANGES contiguous_range<_Range>) {
+            using _String_view = basic_string_view<_CharT>;
+            const auto _Size   = _STD _To_unsigned_like(_RANGES distance(_Rng));
+            if (_Size >= (numeric_limits<typename _String_view::size_type>::max)()) [[unlikely]] {
+                _Throw_format_error("Formatted range is too long.");
+            }
```
Comment (frederick-vs-ja) [inline stl/inc/format:4601]: Does this deserve `if constexpr`?
Also, I believe we can simply spell `size_t` for the size type ([[string.view.template.general]](https://eel.is/c++draft/string.view.template.general)).

File: stl/inc/format [hunk 5]
```diff
@@ -4354,6 +4356,268 @@ public:
...
+    template <_RANGES input_range _Range, class _FormatContext>
+        requires formattable<_RANGES range_reference_t<_Range>, _CharT>
+              && same_as<remove_cvref_t<_RANGES range_reference_t<_Range>>, _Ty>
+    _FormatContext::iterator format(_Range&& _Rng, _FormatContext& _Ctx) const {
+        return _Format(_STD forward<_Range>(_Rng), _Ctx);
+    }
+
+private:
+    template <_RANGES input_range _Range, class _FormatContext>
+    _FormatContext::iterator _Format(_Range&&, _FormatContext&) const; // not defined
```
[the remark this answers is not in the document]
Comment (frederick-vs-ja) [reply stl/inc/format:4523]: Oh, such cases doesn't seem possible. As the iterator type is controlled by the implementation, and is always a `back_insert_iterator`.

Can we `static_assert` or `_STL_INTERNAL_ASSERT` the condition in `range_formatter::format`?

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, this is super awesome! :heart_eyes_cat: I pushed a bunch of nitpicky changes, plus one bugfix and test coverage for it.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I had to push an additional commit to dramatically reduce compiler memory consumption (and improve test throughput) by reducing the number of input ranges being tested.

Memory consumption was a severe problem for the MSVC-internal mirror for a couple of reasons:

* MSVC-internal uses an x86-native compiler, unlike the GitHub PR/CI system which uses x64-hosted x86-targeting,
* MSVC-internal uses a "checked" compiler with assertions enabled, which consumes slightly more memory.

Together, this imposed a 4 GB limit, and `/analyze` configurations (which are extremely memory-hungry) ran out of memory.

With my changes, MSVC-internal compiler memory consumption (as measured by `/d1reportMemorySummary`; that switch works for the public build too) decreased from 2280 MB to 395 MB for plain, and from 4GB+ to 718 MB for `/analyze`.

This also dramatically improved test throughput. On the GitHub side, with my physical machine and the x64-native compiler, test execution time dropped from 114 seconds to 18 seconds, a delicious 6.3x speedup.

**Comment by StephanTLavavej** [comment]
Thanks for this significant step in one of the last remaining C++23 features! :heart_eyes_cat: :chart_with_upwards_trend: :star:

**Pull request merged**